Pr/try redaction (#30)
GoogleCloudPlatform/opentelemetry-collector-builder-sample: deploy/gke/redaction/README.md

```
kubectl delete namespace $OTEL_NAMESPACE
```

> ## Try it out yourself (Optional)

Reviewer: This should go before the cleanup section.  (if someone cleans up, then the try it out won't work).

> This section contains steps that you can follow to actually see redaction you setup earlier in action. Since there is no expected outcome from our kubernetes deployment due to a lack of telemetry data, the redaction processor setup in our collector cofiguration is not applied to anything. In this section we will introduce a source of telemetry data to actually see the redaction in action.  

Reviewer 2: ```suggestion
This section contains steps that you can follow to actually see redaction you setup earlier in action. Since there is no expected outcome from our Kubernetes deployment due to a lack of telemetry data, the redaction processor setup in our collector configuration is not applied to anything. In this section we will introduce a source of telemetry data to actually see the redaction in action.  
```

> This section contains steps that you can follow to actually see redaction you setup earlier in action. Since there is no expected outcome from our kubernetes deployment due to a lack of telemetry data, the redaction processor setup in our collector cofiguration is not applied to anything. In this section we will introduce a source of telemetry data to actually see the redaction in action.  

Reviewer: ```suggestion
Thus far, we've set up an OpenTelemetry collector configured with the redaction processor, but haven't sent it any telemetry data. In this section we will introduce a source of telemetry data to see the redaction processor in action.  
```

> ### Prerequisites 

Reviewer: For each example (e.g. redaction here), there should be one prerequisites section at the top.  Then, users should be able to follow the example from top-to-bottom.  If there are additional overall prerequisites, list them at the top.  If there are additional verification steps, add those as verification step sections in the readme.

Author: Makes sense, I was thinking to remove this heading and just merging its content under "Try it out yourself" heading.

> You should be able to follow all steps mentioned in this README till [Verify the Deployment](#verify-the-deployment) successfully. This ensures that there are no permissions issues and you are able to succeesfully run a GKE cluster and connect to it. 

Reviewer 2: ```suggestion
You should be able to follow all steps mentioned in this README up to [Verify the Deployment](#verify-the-deployment) successfully. This ensures that there are no permissions issues and you are able to successfully run a GKE cluster and connect to it. 
```

### Source of telemetry data

> We would be using JSON files containing telemetry data in OTLP format as the source of telemetry data. A sample file has been provided here - [testdata.json](./otlp-data/testdata.json). The collector running in the cluster would read this file and treat the data in this file as if it were coming from a running application. 

Reviewer 2: ```suggestion
We will be using JSON files containing telemetry data in OTLP format as the source of telemetry data. A sample file has been provided here - [testdata.json](./otlp-data/testdata.json). The collector running in the cluster will read this file and treat the data in this file as if it is coming from a running application. 
```

It sounds a little clearer at this point to assume the reader is actually following the guide (ie, "you will do this step" vs "you would do this step")

> We would be using JSON files containing telemetry data in OTLP format as the source of telemetry data. A sample file has been provided here - [testdata.json](./otlp-data/testdata.json). The collector running in the cluster would read this file and treat the data in this file as if it were coming from a running application. 

Reviewer: ```suggestion
This example uses JSON files containing telemetry data in OTLP format as the source of telemetry data. A sample file has been provided here - [testdata.json](./otlp-data/testdata.json). The collector running in the cluster would read this file and treat the data in this file as if it were coming from a running application. 
```
Wording nit

### Updating the collector configuration to receive data from JSON file

Next, we would update our collector configuration file - [otel-config.yaml](./otel-config.yaml) to add a reciever that is able to receive telemetry data from the added JSON file. 

> Under the `receivers` section in the config file, add the following configuration - 

Reviewer: This step might be error prone for users not familiar to the otel-collector's config structure.  It also doens't have much educational value, IMO (user just copy-pastes).

Can we either:
* Incorporate this into the original otel-config.yaml file, or
* Make a separate otel-config-example.yaml file?

Author: @damemi mentioned that a more comprehensive example code here would be beneficial for the users - since there are multiple receiver sections in the otel config. 

But I like the idea of providing a `otel-config-redaction.yaml` which contains the final expected yaml configuration. We can then also add a few comments within the file and link that file in README for reference.

```
  otlpjsonfile:
    start_at: "beginning"
    include:
      - "/mnt/testdata/*.json"

> 

Reviewer 2: this might not be clear since there are a couple `receivers` sections (one in the component config, one in each metrics/logs/trace pipeline). could you add a sample to show where this fits in?

Author: Will be adding a sample config file that is already configured, and link it in this step for reference.

```
For more details about this particular receiver, check [otlpjsonfilereceiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/otlpjsonfilereceiver).

> ***NOTE:** The path in `include` points to where the file would be mounted within the kubernetes cluster environment and is therefore different from where it is present on your local machine. More information on this in [the next section](#making-the-json-file-available-within-the-cluster).*

Reviewer 2: ```suggestion
***NOTE:** The path in `include` points to where the file would be mounted within the Kubernetes cluster environment and is therefore different from where it is present on your local machine. More information on this in [the next section](#making-the-json-file-available-within-the-cluster).*
```

Next, add this receiver in the traces pipeline, so your trace pipeline looks like - 
```
 traces:
      receivers: [otlp, otlpjsonfile]
      processors: [memory_limiter, batch, resourcedetection/gke, redaction/credit_cards]
      exporters: [googlecloud, logging]
```
 
### Making the JSON file available within the cluster 

> Simply adding the JSON file within the directory does not grant the running cluster the access to this file. In order to get access to this file within the kubernetes environment, we would be mounting this file as a kubernetes `configMap` for a new deployment - 

Reviewer 2: ```suggestion
Simply adding the JSON file within the directory does not grant the running cluster the access to this file. In order to get access to this file within the Kubernetes environment, we will mount this file as a Kubernetes ConfigMap for a new deployment - 
```

Reviewer 2: when it's not clear how a k8s object should be stylized (such as `configMap` vs ConfigMap), I usually fall back to the upstream documentation as a guide: https://kubernetes.io/docs/concepts/configuration/configmap/

please ctrl+f this change throughout these steps

1. Make sure that there are no current deployments by running `kubectl get deployments -n $OTEL_NAMESPACE`. If there are active deployments, delete them using `kubectl delete`. 
2. Since we also updated our collector configuration, we will need to recreate the `otel-config` configMap too. Delete the old configMaps, if any before proceeding. 
    - To delete configMaps, you can use `kubectl delete configmaps <configMap name> -n $OTEL_NAMESPACE`. 

> 3. Recreate `otel-config` configMap with the updated configuration - `kubectl create configmap otel-config --from-file=./otel-config.yaml -n $OTEL_NAMESPACE`. 

Reviewer 2: This section would flow better if each step the user has to do is broken into a code block (three ```) rather than inline (`), eg

3. Recreate `otel-config` configMap with the updated configuration 
```
`kubectl create configmap otel-config --from-file=./otel-config.yaml -n $OTEL_NAMESPACE`. 
```

(same for steps 4,6,7)

> 4. Create a new configMap `otlp-test-data` - this would be used to mount our test data in the kubernetes cluster as a volume - `kubectl create configmap otlp-test-data --from-file=./otlp-data/testdata.json -n $OTEL_NAMESPACE`. 

Reviewer 2: ```suggestion
4. Create a new configMap `otlp-test-data` - this will be used to mount our test data in the Kubernetes cluster as a volume - `kubectl create configmap otlp-test-data --from-file=./otlp-data/testdata.json -n $OTEL_NAMESPACE`. 
```

5. Update the [manifest.yaml](./manifest.yaml) to add `volume` and `volumeMount` configuration for the newly created configMap - `otlp-test-data`. 
   - Update the `spec.volumes` section to add a new `configMap` entry  
        ```
        - configMap:
                name: otlp-test-data
            name: test-data-vol   # This could be changed to anything
        ```
   - Update the `spec.containers.volumeMounts` section to configure a new volumeMount for the added volume. 
        ```
            - name: test-data-vol  # This name should match to the name of volume in spec.volumes

>                 mountPath: /mnt/testdata   # This is the location in the kubernetes environment where the file for configMap will be mounted

Reviewer 2: similar to the sample config comment from earlier, it's not entirely clear where in the pod spec these should go so an expanded code sample will be helpful

        ```
6. Deploy the cluster using the new deployment manifest - `kubectl apply -f manifest.yaml -n $OTEL_NAMESPACE`.
7. Verify if the file is available within the cluster at the mount path specified. This can be done by getting into the running pods within the cluster.
    - `kubectl exec -n $OTEL_NAMESPACE -it <pod_container_name> -- cat /mnt/testdata/testdata.json` - should show the contents of the file

>     - You can get the pod name using `kubectl get pods -n $OTEL_NAMESPACE `

Reviewer 2: Ah I see now why the busybox steps were included ... we aren't really encouraged to use busybox due to its license, so we shouldn't include these steps by default. Our goal is to ship a very minimal docker image and busybox adds a lot of weight to that image too.

Please remove the busybox build steps from the dockerfile. These steps can be moved to their own troubleshooting section (a user shouldn't need to verify the file was mounted if everything is working). That troubleshooting section can show how to modify the Dockerfile to include a shell

Author: I see, makes sense. I can add those steps in troubleshooting
